src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1549,184 +1667,63 @@ private void ExtendWindow(int amount)
-
-            // Invalidate outside of lock to avoid race with HttpPool Dispose()
-            // We should not try to grab pool lock while holding connection lock as on disposing pool,
-            // we could hold pool lock while trying to grab connection lock in Dispose().
-            _pool.InvalidateHttp2Connection(this);
+            if (NetEventSource.Log.IsEnabled()) Trace($"{nameof(abortException)}=={abortException}");
 
+            // The connection has failed, e.g. failed IO or a connection-level protocol error.
             List<Http2Stream> streamsToAbort = new List<Http2Stream>();
-
             lock (SyncObject)
             {
-                // Set _lastStreamId to int.MaxValue to indicate that we are shutting down
-                // and we must assume all active streams have been processed by the server
-                _lastStreamId = int.MaxValue;
-
-                foreach (KeyValuePair<int, Http2Stream> kvp in _httpStreams)
+                if (_abortException is not null)
                 {
-                    int streamId = kvp.Key;
-                    Debug.Assert(streamId == kvp.Value.StreamId);
-
-                    streamsToAbort.Add(kvp.Value);
+                    if (NetEventSource.Log.IsEnabled()) Trace($"Abort called while already aborting. {nameof(abortException)}={abortException}");
+                    return;
                 }
 
-                CheckForShutdown();
-            }
-
-            // Avoid calling OnAbort under the lock, as it may cause the Http2Stream
-            // to call back in to RemoveStream
-            foreach (Http2Stream s in streamsToAbort)
-            {
-                s.OnReset(abortException);
-            }
-        }
-
-        private void StartTerminatingConnection(int lastValidStream, Exception abortException)
-        {
-            Debug.Assert(lastValidStream >= 0);
-
-            // Invalidate outside of lock to avoid race with HttpPool Dispose()
-            // We should not try to grab pool lock while holding connection lock as on disposing pool,
-            // we could hold pool lock while trying to grab connection lock in Dispose().
-            _pool.InvalidateHttp2Connection(this);
-
-            List<Http2Stream> streamsToAbort = new List<Http2Stream>();
-
-            lock (SyncObject)
-            {
-                if (_lastStreamId == -1)
-                {
-                    _lastStreamId = lastValidStream;
-                }
-                else
-                {
-                    // We have already received GOAWAY before
-                    // In this case the smaller valid stream is used
-                    _lastStreamId = Math.Min(_lastStreamId, lastValidStream);
-                }
+                _abortException = abortException;
 
-                if (NetEventSource.Log.IsEnabled()) Trace($"{nameof(lastValidStream)}={lastValidStream}, {nameof(_lastStreamId)}={_lastStreamId}");
+                Shutdown();
```

Comment:
From here to L1711 seems pretty much like the end of `ProcessGoAwayFrame` and  what originally was in `AbortStreams`.
Is there a reason why not have it in a private method?

Reply:
I basically just merged all these functions together because it didn't seem like there was a good reason to have them as separate functions.